### PR TITLE
[Feature] Add FP16 inference support to NMT + Add BoundedBudgetSampler

### PR DESCRIPTION
**Pull request by hymzoque**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- average_checkpoint cli for nmt
- bounded budget sampler, same implementation with fairseq (see https://github.com/pytorch/fairseq/blob/master/fairseq/data/data_utils_fast.pyx)
- add situation eos_id=None when applying beam search in sequence sampler 

cc @dmlc/gluon-nlp-team

### CODE REVIEW

File: scripts/datasets/machine_translation/wmt2014_ende.sh [hunk 1]
```diff
@@ -24,17 +24,21 @@ nlp_preprocess clean_tok_para_corpus --src-lang ${SRC} \
...
                       --tgt-lang ${TGT} \
                       --src-corpus dev.raw.${SRC} \
                       --tgt-corpus dev.raw.${TGT} \
                       --min-num-words 1 \
                       --max-num-words 100 \
+                      --max-ratio 1.5 \
                       --src-save-path dev.tok.${SRC} \
-                      --tgt-save-path dev.tok.${TGT}
+                      --tgt-save-path dev.tok.${TGT} \
+                      --num-process 32
```
Comment (sxjscience) [inline scripts/datasets/machine_translation/wmt2014_ende.sh:41]: Automatically determine the num-process. The user may run the script in other operating system so you should not add the flag.

File: scripts/machine_translation/average_checkpoint.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import mxnet as mx
+
+mx.npx.set_np()
+
+def get_parser():
+    parser = argparse.ArgumentParser(description='Script to average the checkpoints')
+    parser.add_argument('--prefix', type=str, required=True,
+                        help='Prefix of the checkpoint files, e.g. --prefix ckpt_folder/checkpoint')
+    parser.add_argument('--suffix', type=str, default='',
+                        help='Prefix of the checkpoint files, e.g. --suffix .params')
```
Comment (sxjscience) [inline scripts/machine_translation/average_checkpoint.py:11]: Suffix of the checkpoint files.

File: scripts/machine_translation/average_checkpoint.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    parser.add_argument('--range', type=str, required=True,
+                        help='a piece of python code to represent the checkpoint range, '
+                             'e.g. [1,3,7] range(10) range(4000,5000,2) [5,6]+[7,8]')
```
Comment (sxjscience) [inline scripts/machine_translation/average_checkpoint.py:14]: Is it possible to use `nargs='+'`?

File: scripts/machine_translation/average_checkpoint.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    parser.add_argument('--save-path', type=str, required=True,
+                        help='Path of the output file')
+    return parser
+
+def main(args):
+    args.range = eval(args.range)
```
Comment (sxjscience) [inline scripts/machine_translation/average_checkpoint.py:20]: You can not use eval, it's vulnerable.

File: scripts/machine_translation/average_checkpoint.py (hunk 3 above)
Comment (sxjscience) [reply scripts/machine_translation/average_checkpoint.py:14]: So we may use regex in the CLI. For example, `python average_checkpoint.py --checkpoints *.params --begin XX --end XX`.

File: scripts/machine_translation/train_transformer.py [hunk 5]
```diff
@@ -300,6 +314,8 @@ def train(args):
     cfg.defrost()
     cfg.MODEL.src_vocab_size = len(src_vocab)
     cfg.MODEL.tgt_vocab_size = len(tgt_vocab)
+    if args.fp16:
```
Comment (sxjscience) [inline scripts/machine_translation/train_transformer.py:317]: Add a TODO item here or raise an exception that training may not work.

File: scripts/machine_translation/train_transformer.py [hunk 6]
```diff
@@ -325,24 +341,33 @@ def train(args):
...
                              'beta2': 0.98, 'epsilon': 1e-9, 'lr_scheduler': lr_scheduler})
     # Load Data
-    if args.bucket_scheme == 'constant':
-        bucket_scheme = ConstWidthBucket()
-    elif args.bucket_scheme == 'linear':
-        bucket_scheme = LinearWidthBucket()
-    elif args.bucket_scheme == 'exp':
-        bucket_scheme = ExpWidthBucket(bucket_len_step=1.2)
+    if args.sampler == 'BoundedBudgetSampler':
+        train_batch_sampler = BoundedBudgetSampler(lengths=[(ele[2], ele[3]) for ele in data_train],
```
Comment (sxjscience) [inline scripts/machine_translation/train_transformer.py:345]: In Fairseq, they use the max of source and target: https://fairseq.readthedocs.io/en/latest/_modules/fairseq/data/language_pair_dataset.html#LanguagePairDataset.num_tokens

Comment (sxjscience) [reply scripts/machine_translation/train_transformer.py:345]: OK, I find that it's handled in the BoundedBudgetSampler.

File: src/gluonnlp/data/sampler.py [hunk 7]
```diff
@@ -266,8 +266,87 @@ def __len__(self):
...
+    Parameters
+    ----------
+    lengths
+        The length of the sequences in the input data sample.
+    max_tokens
+        max tokens num of each batch
+    max_sentences
+        max sentences num of each batch
+    required_batch_size_multiple
+        require batch size to be a multiple of N (default: 1).
```
Comment (sxjscience) [inline src/gluonnlp/data/sampler.py:282]: Explain the benefit, e.g., better throughput in GPU.

**Comment by sxjscience** [comment]
LGTM overall. One comment about the usage of `eval`. Let's wait for the test.

File: scripts/datasets/machine_translation/wmt2014_ende.sh (hunk 1 above)
Comment (hymzoque) [reply scripts/datasets/machine_translation/wmt2014_ende.sh]: TODO automatically determine the num-process in clean scripts

File: src/gluonnlp/data/sampler.py [hunk 8]
```diff
@@ -266,8 +266,88 @@ def __len__(self):
         return len(self._sorted_ids)
 
 
+class BoundedBudgetSampler(BaseSampler):
```
Comment (sxjscience) [inline src/gluonnlp/data/sampler.py:269]: @szhengac We added this sampler, which is the same as in Fairseq.

File: src/gluonnlp/data/sampler.py [hunk 9]
```diff
@@ -266,8 +266,88 @@ def __len__(self):
...
+    r"""Assign each data sample to bounded budget batches. Samples will be sorted by length before batchfy
+    see https://github.com/pytorch/fairseq/blob/master/fairseq/data/data_utils_fast.pyx
+
+    Parameters
+    ----------
+    lengths
+        The length of the sequences in the input data sample.
+    max_tokens
```
Comment (szhengac) [inline src/gluonnlp/data/sampler.py:277]: change it to max_num_tokens, which is consistent to other naming.

File: src/gluonnlp/data/sampler.py [hunk 10]
```diff
@@ -266,8 +266,88 @@ def __len__(self):
...
+        max tokens num of each batch
+    max_sentences
```
Comment (szhengac) [inline src/gluonnlp/data/sampler.py:279]: similar to the above comment. change it to max_num_sentences.

File: src/gluonnlp/data/sampler.py [hunk 11]
```diff
@@ -266,8 +266,88 @@ def __len__(self):
...
+        for index in self._indices:
+            batch_max_sample_len = max(batch_max_sample_len, self._lengths[index])
+            # try to insert new sample to the batch
+            batch_num_sentences = len(batch) + 1
+            batch_num_tokens = batch_num_sentences * batch_max_sample_len
+            if (self._max_sentences > 0 and batch_num_sentences > self._max_sentences) or \
+               (self._max_tokens > 0 and batch_num_tokens > self._max_tokens):
+                moded_bs = max(
+                    required_batch_size_multiple * (len(batch) // required_batch_size_multiple),
+                    len(batch) % required_batch_size_multiple
```
Comment (szhengac) [inline src/gluonnlp/data/sampler.py:318]: when this can be larger?

Comment (hymzoque) [reply src/gluonnlp/data/sampler.py:318]: this is consistent with
```python
if len(batch) // required_batch_size_multiple == 0:
    moded_bs = len(batch) % required_batch_size_multiple
```

File: src/gluonnlp/cli/average_checkpoint.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    ckpt_paths = [args.checkpoints.replace('*', str(i)) for i in args.range]
+    res = mx.npx.load(ckpt_paths[0])
+    keys = res.keys()
+    for ckpt_path in ckpt_paths[1:]:
+        ckpt = mx.npx.load(ckpt_path)
+        for key in keys:
+            res[key] += ckpt[key]
+    for key in keys:
+        res[key] /= len(args.range)
+    mx.npx.save(args.save_path, res)
```
Comment (sxjscience) [inline src/gluonnlp/cli/average_checkpoint.py:45]: @leezu We decide to use the `save` and `load` functionalities to implement the `average_checkpoint` function, which is similar to that in fairseq: https://github.com/pytorch/fairseq/blob/master/scripts/average_checkpoints.py.

Do you think we will keep supporting the usage like this?

File: tests/test_data_sampler.py [hunk 13]
```diff
@@ -128,3 +128,21 @@ def test_split_sampler_even_size(num_samples, num_parts):
...
+@pytest.mark.parametrize('shuffle', [True, False])
+@pytest.mark.parametrize('seed', [100, None])
+def test_bounded_budget_sampler(seq_lengths, max_num_tokens, max_num_sentences,
+                                required_batch_size_multiple, shuffle, seed):
+    sampler = s.BoundedBudgetSampler(seq_lengths, max_num_tokens, max_num_sentences,
+                                     required_batch_size_multiple, shuffle, seed)
+    print(sampler)
+    total_sampled_ids = []
+    for batch_sample_ids in sampler:
+        total_sampled_ids.extend(batch_sample_ids)
```
Comment (sxjscience) [inline tests/test_data_sampler.py:147]: When shuffle is False, you may test that all indexes are covered.

**Comment by sxjscience** [comment]
@hymzoque Some minor comments from my side. Basically, we need to test whether the BoundedBudgetSampler has covered all samples in the dataset when `shuffle=False`.

File: tests/test_data_sampler.py (hunk 13 above)
Comment (hymzoque) [reply tests/test_data_sampler.py:147]: added 
```python
assert sorted(total_sampled_ids) == list(range(len(total_sampled_ids)))
```
in the test script

File: src/gluonnlp/data/sampler.py (hunk 11 above)
Comment (szhengac) [reply src/gluonnlp/data/sampler.py:319]: Ok. You can add comment there saying moded_bs = len(batch) % required_batch_size_multiple when len(batch) < required_batch_size_multiple
